src/preload/syscall_hook.S

```diff
@@ -572,6 +572,440 @@ SYSCALLHOOK_START(_syscall_hook_trampoline_49_89_ca)
+                        0x06, /* length 6 */                            \
+                        0x0c, /* DW_OP_const4u */                       \
+                        stub_scratch_1 & 0xFF,                          \
+                        (stub_scratch_1 & (0xFF <<  0x8)) >>  0x8,      \
+                        (stub_scratch_1 & (0xFF << 0x10)) >> 0x10,      \
+                        (stub_scratch_1 & (0xFF << 0x18)) >> 0x18,      \
+                        0x06, /* DW_OP_deref */
+        // Move the saving area to the start of scratch_2
+        // Do it in the forward order since we know x8 >= x15
+        ldr     x30, [x8, stub_scratch_2 - preload_thread_locals]
+        str     x30, [x15, stub_scratch_2 - preload_thread_locals]
+        ldr     x30, [x8, stub_scratch_2 - preload_thread_locals + 8]
+        str     x30, [x15, stub_scratch_2 - preload_thread_locals + 8]
+        // Restore x8
+        movk    x8, 0, lsl 16
+        .cfi_restore x8
+        .cfi_escape     0x10, /* DW_CFA_expression */                   \
+                        0x0f, /* x15 */                                 \
+                        0x05, /* length 5 */                            \
+                        0x0c, /* DW_OP_const4u */                       \
+                        stub_scratch_2 & 0xFF,                          \
+                        (stub_scratch_2 & (0xFF <<  0x8)) >>  0x8,      \
+                        (stub_scratch_2 & (0xFF << 0x10)) >> 0x10,      \
+                        (stub_scratch_2 & (0xFF << 0x18)) >> 0x18
+        .cfi_escape     0x10, /* DW_CFA_expression */                           \
+                        0x1e, /* x30 */                                         \
+                        0x05, /* length 5 */                                    \
+                        0x0c, /* DW_OP_const4u */                               \
+                        (stub_scratch_2 + 8) & 0xFF,                            \
+                        ((stub_scratch_2 + 8) & (0xFF <<  0x8)) >>  0x8,        \
+                        ((stub_scratch_2 + 8) & (0xFF << 0x10)) >> 0x10,        \
+                        ((stub_scratch_2 + 8) & (0xFF << 0x18)) >> 0x18
+
+        cmp     x8, 0xdc // SYS_clone
+        .cfi_remember_state
+        b.eq    .Lfallback_rawsyscall
+
+        ldr     w30, [x15, alt_stack_nesting_level - preload_thread_locals]
+        cmp     w30, 0
+        add     w30, w30, 1
+        str     w30, [x15, alt_stack_nesting_level - preload_thread_locals]
+
+        b.ne    .Lnest_syscall_hook_trampoline_raw
+        ldr     x30, [x15, syscallbuf_stub_alt_stack - preload_thread_locals]
+        sub     x30, x30, 48
+        b       .Lstackset_syscall_hook_trampoline_raw
+.Lnest_syscall_hook_trampoline_raw:
+        sub     x30, sp, 48
+.Lstackset_syscall_hook_trampoline_raw:
+        // Now x30 points to the new stack with 48 bytes of space allocated
+        .cfi_def_cfa x30, 48
```

Comment:
This looks like the CFA changes halfway through the function. That will confuse the unwinder. You need to make sure that the CFA has the same at every instruction in the function.

Reply:
I checked that GDB is fine with this as least when I single instruction through the function.
I am wondering what the CFA should be though... The aadwarf64 defined it as the SP before the function call. I can do that fairly easily (need to figure out how to encode it.) but then I guess I just can't use that to save/restore any registers since nothing will be stored relative to that. Alternatively, since at least gdb seems fairly happy with my explicit sp restoring unwind info, I could define CFA as the actual top of the stack frame in this function. This is what I'm doing here, but it's near impossible to encode the CFA before reaching here (unless I essentially write this stack pointer bumping logic again in dwarf expression....). Which way do you think is better?

Reply:
Use the sp when entering the function as the CFA. Yes, it'll make encoding the saved registers more annoying, but I think it's the most correct. Otherwise you'd have to replicate the CFA computation logic in DWARF at the start of the function, which is ... Unfun.